File: 2015/5/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

bool check_good_string_part1(char *line, int len);
bool check_good_string_part2(char *line, int len);

int main(int argc, char **argv)
{
    if (argc < 2)
    {
        fprintf(stderr, "Usage: %s <input.txt>\n", argv[0]);
        return EXIT_FAILURE;
    }

    FILE *file = fopen(argv[1], "r");

    if (file == NULL)
    {
        fprintf(stderr, "Error: Could not open file %s\n", argv[1]);
        return EXIT_FAILURE;
    }

    char *line = NULL;
    size_t len = 0;
    int part1_count = 0;
    int part2_count = 0;

    while (getline(&line, &len, file) != -1)
    {
        // Remove newline character
        line[strcspn(line, "\n")] = 0;
        len = strlen(line);

        if (check_good_string_part1(line, len))
            part1_count++;

        if (check_good_string_part2(line, len))
            part2_count++;
    }

    printf("%d\n", part1_count);
    printf("%d\n", part2_count);
    free(line);
    fclose(file);
    return EXIT_SUCCESS;
}

bool check_good_string_part1(char *line, int len)
{
    int vowels = 0;
    bool double_letter = false;
    bool bad_string = false;

    for (int i = 0; i < len; i++)
    {
        if (line[i] == 'a' || line[i] == 'e' || line[i] == 'i' || line[i] == 'o' || line[i] == 'u')
            vowels++;

        if (i > 0 && line[i] == line[i - 1])
            double_letter = true;

        if (i > 0 && ((line[i] == 'b' && line[i - 1] == 'a') || (line[i] == 'd' && line[i - 1] == 'c') || (line[i] == 'q' && line[i - 1] == 'p') || (line[i] == 'y' && line[i - 1] == 'x')))
            bad_string = true;
    }

    return vowels >= 3 && double_letter && !bad_string;
}

bool check_good_string_part2(char *line, int len)
{
    bool pair = false;
    bool repeat = false;

    for (int i = 0; i < len - 1; i++)
    {
        for (int j = i + 2; j < len - 1; j++)
        {
            if (line[i] == line[j] && line[i + 1] == line[j + 1])
                pair = true;
        }

        if (i < len - 2 && line[i] == line[i + 2])
            repeat = true;
    }

    return pair && repeat;
}

File: 2015/typescript/5/main.ts
import { PathLike, existsSync, readFileSync } from "fs";

function check_good_string_part1(s: string): boolean {
  let vowels: number = 0;
  let double_letter: boolean = false;
  let bad_strings: boolean = false;

  for (let i = 0; i < s.length; i++) {
    if ("aeiou".includes(s[i])) {
      vowels++;
    }

    if (i > 0 && s[i] === s[i - 1]) {
      double_letter = true;
    }

    if (["ab", "cd", "pq", "xy"].includes(s.slice(i, i + 2))) {
      bad_strings = true;
    }
  }

  return vowels >= 3 && double_letter && !bad_strings;
}

function check_good_string_part2(s: string): boolean {
  let double_pair: boolean = false;
  let repeat_with_gap: boolean = false;

  for (let i = 0; i < s.length - 1; i++) {
    if (s.slice(i + 2).includes(s.slice(i, i + 2))) {
      double_pair = true;
    }

    if (s[i] === s[i + 2]) {
      repeat_with_gap = true;
    }
  }

  return double_pair && repeat_with_gap;
}

function main() {
  // get command line arguments
  const args: string[] = process.argv.slice(1);

  if (args.length < 2) {
    console.log("Usage: node main.js <input.txt>");
    return;
  }

  const filename: string = args[1];

  if (!existsSync(filename as PathLike)) {
    console.log(`File ${filename} does not exist`);
    return;
  }

  const data: string = readFileSync(filename as PathLike, "utf8");

  const lines: string[] = data.split("\n");

  let part1_count: number = 0;
  let part2_count: number = 0;

  for (let line of lines) {
    if (check_good_string_part1(line)) {
      part1_count++;
    }

    if (check_good_string_part2(line)) {
      part2_count++;
    }
  }

  console.log("Part 1:", part1_count);
  console.log("Part 2:", part2_count);
}

main();
